tollway/utils.py

```diff
@@ -1,34 +1,67 @@
 import json
 from pathlib import Path
-from typing import TypedDict
+from typing import Mapping, TypedDict, Union
 
+from decouple import config
 from google.cloud import pubsub_v1
-from google.pubsub_v1 import PubsubMessage
-
-from tollway.constants import PROJECT_ID, TOPIC_ID
+from google.oauth2 import service_account
 
 
 class EventsLog(TypedDict):
     late_events: dict[str, list[str]]
     past_events: list[dict[str, str]]
-    all_events: list[dict[str, str]]
+    all_events: list[Mapping[str, Union[str, bool]]]
 
 
 def get_topic(pubsub: bool) -> tuple:
     if pubsub:
-        project_id = PROJECT_ID
-        topic_id = TOPIC_ID
-        publisher = pubsub_v1.PublisherClient()
-        topic_path = publisher.topic_path(project=project_id, topic=topic_id)
+        project = config("PROJECT_ID")
+        topic = config("TOPIC_ID")
+        credentials = service_account.Credentials.from_service_account_file(config("PUBSUB_SERVICE_ACCOUNT"))
+        publisher = pubsub_v1.PublisherClient(credentials=credentials)
+        topic_path = publisher.topic_path(project=project, topic=topic)
         return publisher, topic_path
     return None, None
 
 
-def encode_message(message: dict) -> PubsubMessage:
-    return PubsubMessage(json.dumps(message).encode("utf-8"))
+def encode_message(message: dict) -> bytes:
+    return json.dumps(message).encode("utf-8")
 
 
-def write_to_file(filename: str, events_log: list[dict]):
+def write_to_file(filename: str, events_log: list[Mapping[str, Union[str, bool]]]):
     path = Path().absolute()
     with open(path / filename, mode="a") as file:
         json.dump(obj=events_log, fp=file, indent=1)
+
+
+def get_envs() -> dict:
+    envs = {}
+
+    duplicate_rate = config("DUPLICATE_RATE", default=50)
+    all_events_count = config("ALL_EVENTS_COUNT", default=250)
+    late_seconds_rate = config("LATE_SECONDS_RATE", default=10)
+    late_minutes_rate = config("LATE_MINUTES_RATE", default=20)
+    late_hours_rate = config("LATE_HOURS_RATE", default=30)
+    late_days_rate = config("LATE_DAYS_RATE", default=100)
+
+    if duplicate_rate == "":
+        duplicate_rate = 50
```

Comment:
what about making those one-liners like:

`duplicate_rate = duplicate_rate or 50`

Reply:
also, I think this can be done straight in `config` above? why the extra step here?

Reply:
as you might've seen, I reworked this is the other PR. If there's no .env, then the use of `config` and `default` are working. If there is a .env file but only some of the environment variables are defined like so

```
PROJECT_ID=project-12345
TOPIC_ID=tollway
PUBSUB_SERVICE_ACCOUNT=/Users/bengriffith/Projects/tollway-traffic/service_account/pubsub.json
GOOGLE_REGION=us-central1
LATE_SECONDS_RATE=
LATE_MINUTES_RATE=
LATE_HOURS_RATE=
LATE_DAYS_RATE=
DUPLICATE_RATE=
ALL_EVENTS_COUNT=
```

then I was running into a situation where `config` was returning an empty string, not default value.

Reply:
I tested this and yes, default does not kick in because as far as python-decouple is concerned, you intended to give it an empty string.

ChatGPT suggests this wrapper:

```
from decouple import config

def get_config_with_default(key, default_value):
    # Get the value using config
    value = config(key, default=default_value)
    # Return the default value if the actual value is None or an empty string
    return default_value if value.strip() == '' else value

# Usage
LATE_SECONDS_RATE = get_config_with_default('LATE_SECONDS_RATE', 'your_default_value')
```

But I wonder if we should just assume that the user needs to set those, I mean they either A. omit them and the default kicks in, B. they set them completely as key=value. If they set them as key= it's actually a "user error"?

What do you think?

Reply:
and therefor ideally we clean up the extra 

```
    if duplicate_rate == "":
        duplicate_rate = 50
....
```

checks ...

Reply:
I like the suggested wrapper and while I do like the idea of assuming the user needs to set those I think just go with the assumption that the user is not going to create `.env` unless they want to use pubsub functionality or they want more control in the rate at which duplicate or late messages are generated.

Reply:
ok so will you change this to add the wrapper function to have all bases covered? or issue and defer?